Mathlib/Data/Tree.lean

```diff
@@ -0,0 +1,180 @@
+  | PosNum.bit0 n, node _a t₁ _t₂ => t₁.get n
+  | PosNum.bit1 n, node _a _t₁ t₂ => t₂.get n
+#align tree.get Tree.get
+
+/-- Retrieves an element from the tree, or the provided default value
+if the index is invalid. See `Tree.get`. -/
+def getOrElse (n : PosNum) (t : Tree α) (v : α) : α :=
+  (t.get n).getD v
+#align tree.get_or_else Tree.getOrElse
+
+/-- Apply a function to each value in the tree.  This is the `map` function for the `tree` functor.
+TODO: implement `traversable tree`. -/
+def map {β} (f : α → β) : Tree α → Tree β
+  | nil => nil
+  | node a l r => node (f a) (map f l) (map f r)
+#align tree.map Tree.map
+
+/-- The number of internal nodes (i.e. not including leaves) of a binary tree -/
+@[simp]
+def numNodes : Tree α → ℕ
+  | nil => 0
+  | node _ a b => a.numNodes + b.numNodes + 1
+#align tree.num_nodes Tree.numNodes
+
+/-- The number of leaves of a binary tree -/
+@[simp]
+def numLeaves : Tree α → ℕ
+  | nil => 1
+  | node _ a b => a.numLeaves + b.numLeaves
+#align tree.num_leaves Tree.numLeaves
+
+/-- The height - length of the longest path from the root - of a binary tree -/
+@[simp]
+def height : Tree α → ℕ
+  | nil => 0
+  | node _ a b => max a.height b.height + 1
+#align tree.height Tree.height
+
+theorem numLeaves_eq_numNodes_succ (x : Tree α) : x.numLeaves = x.numNodes + 1 := by
+  induction x <;> simp [*, Nat.add_comm, Nat.add_assoc, Nat.add_left_comm]
+#align tree.num_leaves_eq_num_nodes_succ Tree.numLeaves_eq_numNodes_succ
+
+theorem numLeaves_pos (x : Tree α) : 0 < x.numLeaves := by
+  rw [numLeaves_eq_numNodes_succ]
+  exact x.numNodes.zero_lt_succ
+#align tree.num_leaves_pos Tree.numLeaves_pos
+
+theorem height_le_numNodes : ∀ x : Tree α, x.height ≤ x.numNodes
+  | nil => le_rfl
+  | node _ a b =>
+    Nat.succ_le_succ
+      (max_le (_root_.trans a.height_le_numNodes <| a.numNodes.le_add_right _)
+        (_root_.trans b.height_le_numNodes <| b.numNodes.le_add_left _))
+#align tree.height_le_num_nodes Tree.height_le_numNodes
+
+/-- The left child of the tree, or `nil` if the tree is `nil` -/
+@[simp]
+def left : Tree α → Tree α
+  | nil => nil
+  | node _ l _r => l
+#align tree.left Tree.left
+
+/-- The right child of the tree, or `nil` if the tree is `nil` -/
+@[simp]
+def right : Tree α → Tree α
+  | nil => nil
+  | node _ _l r => r
+#align tree.right Tree.right
+
+-- mathport name: «expr △ »
+-- Notation for making a node with `Unit` data
+scoped infixr:65 " △ " => Tree.node ()
+
+/-- Recursion on `Tree Unit`; allows for a better `induction` which does not have to worry
+  about the element of type `α = Unit` -/
+@[elab_as_elim]
+def unitRecOn {motive : Tree Unit → Sort _} (t : Tree Unit) (base : motive nil)
+    (ind : ∀ x y, motive x → motive y → motive (x △ y)) : motive t := by
+  -- Porting note: code generator does not support recursor 'Tree.recOn'
+  match t with
+  | nil => exact base
+  | node _a l r => exact ind l r (unitRecOn l base ind) (unitRecOn r base ind)
```

Comment:
Instead of making this change, can you copy the approach of `Nat.rec_eq_recC` for `Tree.recOn`?

Reply:
Tried

```lean
@[csimp]
lemma recOn_Unit_eq_recOnC : @Tree.recOn Unit = @Tree.recOnC := by
```

got

```
invalid 'csimp' theorem, only constant replacement theorems (e.g., `@f = @g`) are currently supported.
```

Reply:
The universe parameters have to be in exactly the same order for it to work

Reply:
I thought that it is because that this function is only for `Tree Unit` (`def unitRecOn {motive : Tree Unit → Sort _}`) while `csimp` only work on `@Tree.recOn = @...` but not `@Tree.recOn Unit = ...`?

Reply:
Oh, sorry; I meant that you should define recC in the general case, not just for Unit

Reply:
Done. Hopefully I did this right.